app/models/generic_object_definition.rb

```diff
@@ -8,44 +8,94 @@ class GenericObjectDefinition < ApplicationRecord
+                    :normalize_property_associations
+
+  before_destroy    :check_not_in_use
 
   def create_object(options)
     GenericObject.create!({:generic_object_definition => self}.merge(options))
   end
 
-  def defined_property_attributes
-    properties[:attributes]
+  FEATURES.each do |feature|
+    define_method("defined_property_#{feature}s") do
+      return errors[:properties] if properties_changed? && !valid?
+      properties["#{feature}s".to_sym]
+    end
+
+    define_method("property_#{feature}_defined?") do |attr|
+      send("defined_property_#{feature}s").try(:key?, attr.to_s)
+    end
   end
 
-  def property_attribute_defined?(attr)
-    defined_property_attributes.try(:key?, attr.to_s)
+  def property_defined?(attr)
+    property_attribute_defined?(attr) || property_association_defined?(attr)
   end
 
-  def properties=(props)
-    props = props.symbolize_keys
-    if props.key?(:attributes)
-      props[:attributes] = props[:attributes].each_with_object({}) do |(name, type), hash|
-        raise ArgumentError, "#{type} is not a recognized type" unless TYPE_MAP.key?(type.to_sym)
-        hash[name.to_s] = type.to_sym
-      end
-    end
-    super
+  def property_getter(attr, val)
+    return type_cast(attr, val) if property_attribute_defined?(attr)
+    return get_associations(attr, val) if property_association_defined?(attr)
   end
 
-  def type_cast(attr_name, value)
-    TYPE_MAP.fetch(defined_property_attributes[attr_name]).cast(value)
+  def type_cast(attr, value)
+    TYPE_MAP.fetch(defined_property_attributes[attr]).cast(value)
   end
 
   private
 
+  def get_associations(attr, values)
+    defined_property_associations[attr].constantize.where(:id => values).to_a
+  end
+
+  def normalize_property_attributes
+    props = properties.symbolize_keys
+
+    properties[:attributes] = props[:attributes].each_with_object({}) do |(name, type), hash|
+      hash[name.to_s] = type.to_sym
+    end if props[:attributes]
+  end
+
+  def normalize_property_associations
+    props = properties.symbolize_keys
+
+    properties[:associations] = props[:associations].each_with_object({}) do |(name, type), hash|
+      hash[name.to_s] = type.to_s.classify
+    end if props[:associations]
+  end
+
+  def validate_property_attributes
+    properties[:attributes].each do |name, type|
+      errors[:properties] << "attribute [#{name}] is not of a recognized type: [#{type}]" unless TYPE_MAP.key?(type.to_sym)
+    end if properties[:attributes]
+  end
+
+  def validate_property_associations
+    properties[:associations].each do |name, klass|
+      begin
+        klass.constantize
+      rescue NameError
+        errors[:properties] << "association [#{name}] is not of a valid model: [#{klass}]"
+      end
+    end if properties[:associations]
+  end
+
+  def validate_property_name_unique
+    common = (properties[:attributes].try(:keys) || []) & (properties[:associations].try(:keys) || [])
```

Comment:
It is a pain to test whether `properties[:associations]` or `properties[:associations]` exist everywhere. You can try to add a default value to `[]` to `properties` using `default_proc` call back